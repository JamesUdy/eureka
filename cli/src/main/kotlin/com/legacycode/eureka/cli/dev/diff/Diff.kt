package com.legacycode.eureka.cli.dev.diff

import com.legacycode.eureka.viz.edgebundling.EdgeBundlingGraph
import com.legacycode.eureka.viz.edgebundling.EdgeBundlingGraph.Link
import com.legacycode.eureka.viz.edgebundling.EdgeBundlingGraph.Node

data class Diff(val missing: Missing) {
  companion object {
    fun of(
      baseline: EdgeBundlingGraph,
      implementation: EdgeBundlingGraph,
    ): Diff {
      val missingNodes = baseline.nodes - implementation.nodes.toSet()
      val missingLinks = baseline.links - implementation.links.toSet()
      return Diff(Missing(missingNodes, missingLinks))
    }
  }
}

data class Missing(
  val nodes: List<Node>,
  val links: List<Link>,
)

val Diff.report: String
  get() {
    val (nodes, links) = missing
    return if (nodes.isEmpty()) {
      "✅ All good, no differences found."
    } else {
      """
        |❌ Missing nodes (${nodes.count()}):
        |${nodes.joinToString(separator = System.lineSeparator()) { "  - (group: ${it.group}) ${it.id}" }}
        |""".trimMargin() +
        """
        |❌ Missing links (${links.count()}):
        |${links.joinToString(separator = System.lineSeparator()) { "  - ${it.source} -> ${it.target}" }}
        |""".trimMargin()
    }
  }
