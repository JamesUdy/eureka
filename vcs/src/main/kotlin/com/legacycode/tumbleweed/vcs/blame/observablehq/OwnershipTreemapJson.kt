package com.legacycode.tumbleweed.vcs.blame.observablehq

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.kotlinModule
import com.legacycode.tumbleweed.vcs.blame.BlameResult

data class Ownership(
  val name: String,
  val size: Int,
  val percentage: Double,
)

data class OwnershipTreemapJson(
  val name: String,
  val children: List<Ownership>,
) {
  companion object {
    fun from(blameResult: BlameResult): OwnershipTreemapJson {
      val ownershipByEmail = blameResult.byEmail()
      val ownerships = ownershipByEmail.map { (email, blameLines) ->
        val percentage = calculatePercentage(blameLines.size, blameResult.lines.size)
        Ownership(email.address, blameLines.size, percentage)
      }
      return OwnershipTreemapJson(blameResult.repoFile.name, ownerships)
    }

    private fun calculatePercentage(
      linesByOwner: Int,
      totalLines: Int,
    ): Double {
      val percentage = linesByOwner / totalLines.toDouble() * 100
      return String.format("%.2f", percentage).toDouble()
    }
  }

  fun toJson(): String {
    return jacksonObjectMapper()
      .registerModule(kotlinModule())
      .writeValueAsString(this)
  }
}
