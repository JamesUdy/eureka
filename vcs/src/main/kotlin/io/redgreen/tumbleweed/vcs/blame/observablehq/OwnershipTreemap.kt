package io.redgreen.tumbleweed.vcs.blame.observablehq

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.kotlinModule
import io.redgreen.tumbleweed.vcs.blame.BlameResult

data class Ownership(
  val name: String,
  val size: Int,
  val percentage: Double,
)

data class OwnershipTreemap(
  val name: String,
  val children: List<Ownership>,
) {
  companion object {
    fun from(blameResult: BlameResult): OwnershipTreemap {
      val ownershipByEmail = blameResult.byEmail()
      val ownerships = ownershipByEmail.map { (email, blameLines) ->
        val percentage = calculatePercentage(blameLines.size, blameResult.lines.size)
        Ownership(email.address, blameLines.size, percentage)
      }
      return OwnershipTreemap(blameResult.repoFile.name, ownerships)
    }

    private fun calculatePercentage(
      linesByOwner: Int,
      totalLines: Int,
    ): Double {
      val percentage = linesByOwner / totalLines.toDouble() * 100
      return String.format("%.2f", percentage).toDouble()
    }
  }

  fun toJson(): String {
    return jacksonObjectMapper()
      .registerModule(kotlinModule())
      .writeValueAsString(this)
  }
}
