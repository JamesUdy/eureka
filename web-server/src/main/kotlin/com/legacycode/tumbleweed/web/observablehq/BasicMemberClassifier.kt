package com.legacycode.tumbleweed.web.observablehq

import com.legacycode.tumbleweed.Field
import com.legacycode.tumbleweed.Member
import com.legacycode.tumbleweed.Method

class BasicMemberClassifier : MemberClassifier {
  override fun groupOf(member: Member): Int {
    return when (member) {
      is Field -> 1
      is Method -> 2
    }
  }
}
