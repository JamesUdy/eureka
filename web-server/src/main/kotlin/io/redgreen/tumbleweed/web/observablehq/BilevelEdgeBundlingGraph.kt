package io.redgreen.tumbleweed.web.observablehq

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper

data class BilevelEdgeBundlingGraph(
  val nodes: List<Node>,
  val links: List<Link>,
) {
  val json: String
    get() {
      return jacksonObjectMapper()
        .writerWithDefaultPrettyPrinter()
        .writeValueAsString(this)
    }

  data class Node(
    val id: String,
    val group: Int,
  )

  data class Link(
    val source: String,
    val target: String,
    val value: Int = 1,
  )
}
