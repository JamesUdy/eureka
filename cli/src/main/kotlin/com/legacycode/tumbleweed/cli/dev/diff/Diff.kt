package com.legacycode.tumbleweed.cli.dev.diff

import com.legacycode.tumbleweed.web.observablehq.BilevelEdgeBundlingGraph
import com.legacycode.tumbleweed.web.observablehq.BilevelEdgeBundlingGraph.Link
import com.legacycode.tumbleweed.web.observablehq.BilevelEdgeBundlingGraph.Node

data class Diff(val missing: Missing) {
  companion object {
    fun of(
      baseline: BilevelEdgeBundlingGraph,
      implementation: BilevelEdgeBundlingGraph,
    ): Diff {
      val missingNodes = baseline.nodes - implementation.nodes.toSet()
      val missingLinks = baseline.links - implementation.links.toSet()
      return Diff(Missing(missingNodes, missingLinks))
    }
  }
}

data class Missing(
  val nodes: List<Node>,
  val links: List<Link>,
)

val Diff.report: String
  get() {
    val (nodes, links) = missing
    return if (nodes.isEmpty()) {
      "✅ All good, no differences found."
    } else {
      """
        |❌ Missing nodes (${nodes.count()}):
        |${nodes.joinToString(separator = System.lineSeparator()) { "  - (group: ${it.group}) ${it.id}" }}
        |""".trimMargin() +
      """
        |❌ Missing links (${links.count()}):
        |${links.joinToString(separator = System.lineSeparator()) { "  - ${it.source} -> ${it.target}" }}
        |""".trimMargin()
    }
  }
