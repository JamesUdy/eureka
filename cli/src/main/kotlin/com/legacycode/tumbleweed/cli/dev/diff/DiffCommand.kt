package com.legacycode.tumbleweed.cli.dev.diff

import com.legacycode.tumbleweed.web.observablehq.BilevelEdgeBundlingGraph
import java.io.File
import picocli.CommandLine.Command
import picocli.CommandLine.Option

@Command(
  name = "diff",
  description = ["(dev) compare two ObservableHQ JSON files"],
  hidden = true,
)
class DiffCommand : Runnable {
  @Option(
    names = ["-b", "--baseline"],
    description = ["path to the baseline JSON file"],
    required = true,
  )
  lateinit var baselineJson: File

  @Option(
    names = ["-i", "--implementation"],
    description = ["path to the implementation JSON file"],
    required = true,
  )
  lateinit var implementationJson: File

  override fun run() {
    val baseline = BilevelEdgeBundlingGraph.fromJson(baselineJson.readText())
    val implementation = BilevelEdgeBundlingGraph.fromJson(implementationJson.readText())

    val diff = Diff.of(baseline, implementation)

    println(diff.report)
  }
}
