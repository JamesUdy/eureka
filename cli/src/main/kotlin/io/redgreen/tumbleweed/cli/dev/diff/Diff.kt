package io.redgreen.tumbleweed.cli.dev.diff

import io.redgreen.tumbleweed.web.observablehq.BilevelEdgeBundlingGraph
import io.redgreen.tumbleweed.web.observablehq.BilevelEdgeBundlingGraph.Node

data class Diff(val missing: Missing) {
  companion object {
    fun of(
      baseline: BilevelEdgeBundlingGraph,
      implementation: BilevelEdgeBundlingGraph,
    ): Diff {
      val missingNodes = baseline.nodes - implementation.nodes.toSet()
      return Diff(Missing(missingNodes))
    }
  }
}

data class Missing(
  val nodes: List<Node>,
)

val Diff.printable: String
  get() {
    val missingNodes = missing.nodes
    return if (missingNodes.isEmpty()) {
      "✅ All good, no differences found."
    } else {
      """
          |❌ Differences found:
          |${missingNodes.joinToString(separator = System.lineSeparator()) { "  - (${it.group}) ${it.id}" }}
          |""".trimMargin()
    }
  }
