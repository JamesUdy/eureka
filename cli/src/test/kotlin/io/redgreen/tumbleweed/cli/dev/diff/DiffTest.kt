@file:Suppress("JUnitMalformedDeclaration")

package io.redgreen.tumbleweed.cli.dev.diff

import io.redgreen.tumbleweed.cli.dev.diff.EdgeBundlingGraphExtension.*
import io.redgreen.tumbleweed.web.observablehq.BilevelEdgeBundlingGraph
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(EdgeBundlingGraphExtension::class)
class DiffTest {
  @Test
  fun `it returns an empty list when the baseline and implementation are same`(
    @Graph("01-baseline.csv") baseline: BilevelEdgeBundlingGraph,
    @Graph("01-implementation.csv") implementation: BilevelEdgeBundlingGraph,
  ) {
    // when
    val diff = Diff.of(baseline, implementation)

    // then
    Approvals.verify(diff.report)
  }

  @Test
  fun `it returns a list of missing nodes when the implementation is empty`(
    @Graph("02-baseline.csv") baseline: BilevelEdgeBundlingGraph,
    @Graph("02-implementation.csv") implementation: BilevelEdgeBundlingGraph,
  ) {
    // when
    val diff = Diff.of(baseline, implementation)

    // then
    Approvals.verify(diff.report)
  }

  @Test
  fun `it returns a list of missing nodes when the implementation is missing some nodes`(
    @Graph("03-baseline.csv") baseline: BilevelEdgeBundlingGraph,
    @Graph("03-implementation.csv") implementation: BilevelEdgeBundlingGraph,
  ) {
    // when
    val diff = Diff.of(baseline, implementation)

    // then
    Approvals.verify(diff.report)
  }
}
