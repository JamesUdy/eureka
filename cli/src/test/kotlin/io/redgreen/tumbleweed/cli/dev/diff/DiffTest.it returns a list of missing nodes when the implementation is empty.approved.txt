❌ Differences found:
  - (1) String name
  - (2) void getName()
  - (2) void setName(String)
